### PR TITLE
feature(zc): add bcq algorithm

### PR DESCRIPTION
**Pull request by Super1ce**
## Description
Add bcq algorithm

## TODO
Add algorithm and config

BCQ paper :https://arxiv.org/pdf/1812.02900.pdf
reference implementation :https://github.com/sfujim/BCQ

## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

**Comment by PaParaZz1** [comment]
add paper link and a training curve figure in PR

File: ding/example/bcq.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+from ding.envs import DingEnvWrapper, BaseEnvManagerV2
+from ding.data import create_dataset
+from ding.config import compile_config
+from ding.framework import task, ding_init
+from ding.framework.context import OfflineRLContext
+from ding.framework.middleware import interaction_evaluator, trainer, CkptSaver, offline_data_fetcher, offline_logger
+from ding.utils import set_pkg_seed
+from dizoo.d4rl.envs import D4RLEnv
+from dizoo.d4rl.config.halfcheetah_medium_bcq_config import main_config, create_config
+# from dizoo.d4rl.config.halfcheetah_medium_expert_edac_config import main_config,create_config
```
Comment (zjowowen) [inline ding/example/bcq.py:14]: remove unused codes.

File: ding/model/template/bcq.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            obs_shape: Union[int, SequenceType],
+            action_shape: Union[int, SequenceType, EasyDict],
+            actor_head_hidden_size: int = 64,
+            critic_head_hidden_size: int = 64,
+            activation: Optional[nn.Module] = nn.ReLU(),
+            norm_type: Optional[str] = None,
+            vae_hidden_dims: List = [750, 750],
+            phi: float = 0.05
+    ) -> None:
+        super(BCQ, self).__init__()
```
Comment (zjowowen) [inline ding/model/template/bcq.py:28]: add note for arguments.

File: ding/model/template/bcq.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        self.vae = VanillaVAE(action_shape, obs_shape, action_shape * 2, vae_hidden_dims)
+
+    def forward(self, inputs: Union[torch.Tensor, Dict[str, torch.Tensor]], mode: str) -> Dict[str, torch.Tensor]:
+        """
+        Overview:
+            The unique execution (forward) method of QAC method, and one can indicate different modes to implement \
+            different computation graph, including ``compute_actor`` and ``compute_critic`` in QAC.
+        Mode compute_actor:
+            Arguments:
+                - inputs (:obj:`torch.Tensor`): Observation data, defaults to tensor.
```
Comment (zjowowen) [inline ding/model/template/bcq.py:65]: arguments object type is Union[torch.Tensor, Dict[str, torch.Tensor]] for inputs?

File: dizoo/d4rl/entry/d4rl_bcq_main.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+from ding.entry import serial_pipeline_offline
+from ding.config import read_config
+from pathlib import Path
+
+
+def train(args):
+    # launch from anywhere
+    config = Path(__file__).absolute().parent.parent / 'config' / args.config
```
Comment (zjowowen) [inline dizoo/d4rl/entry/d4rl_bcq_main.py:8]: import from dizoo config file directly?

Comment (Super1ce) [reply dizoo/d4rl/entry/d4rl_bcq_main.py:8]: yes, I copy this from other d4rl_*_main.py

**Review by PaParaZz1 — APPROVED** [review]

**Review by PaParaZz1 — APPROVED** [review]

**Review by PaParaZz1 — APPROVED** [review]

**Review by PaParaZz1 — APPROVED** [review]

**Pull request merged**